libavcodec/libaribcaption.c

```diff
@@ -0,0 +1,1141 @@
+        break;
+
+    case SUBTITLE_NONE:
+    default:
+        status = 0;
+    }
+
+    if (status < 0) {
+        av_log(ctx, AV_LOG_ERROR, "Failed to set Subtitle: %s\n",
+               av_err2str(status));
+        aribcc_caption_cleanup(&ctx->caption);
+        return status;
+    }
+    if (status > 0) {
+        *got_sub_ptr = 1;
+        if (ctx->avpkt->pts != AV_NOPTS_VALUE)
+            sub->pts = av_rescale_q(ctx->avpkt->pts,
+                                    ctx->time_base, AV_TIME_BASE_Q);
+        if (ctx->caption.wait_duration == ARIBCC_DURATION_INDEFINITE)
+            sub->end_display_time = UINT32_MAX;
+        else
+            sub->end_display_time = (uint32_t)ctx->caption.wait_duration;
+    }
+
+    aribcc_caption_cleanup(&ctx->caption);
+    return avpkt->size;
+}
+
+static void aribcaption_flush(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+
+    if (!(avctx->flags2 & AV_CODEC_FLAG2_RO_FLUSH_NOOP)) {
+        if (ctx->decoder)
+            aribcc_decoder_flush(ctx->decoder);
+        if (ctx->renderer)
+            aribcc_renderer_flush(ctx->renderer);
+        ctx->readorder = 0;
+    }
+}
+
+static int aribcaption_close(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+
+    aribcaption_flush(avctx);
+
+    if (ctx->clut)
+        av_freep(&ctx->clut);
+    if (ctx->renderer) {
+        aribcc_renderer_free(ctx->renderer);
+        ctx->renderer = NULL;
+    }
+    if (ctx->decoder) {
+        aribcc_decoder_free(ctx->decoder);
+        ctx->decoder = NULL;
+    }
+    if (ctx->context) {
+        aribcc_context_free(ctx->context);
+        ctx->context = NULL;
+     }
+
+    return 0;
+}
+
+static int aribcaption_init(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+    aribcc_profile_t profile;
+    aribcc_encoding_scheme_t encode_scheme;
+    int ret = 0;
+
+    ctx->avctx = avctx;
+
+    switch (avctx->profile) {
+    case FF_PROFILE_ARIB_PROFILE_A:
+        profile = ARIBCC_PROFILE_A;
+        /* assume 960x540 at initial state */
+        ctx->plane_width = 960;
+        ctx->plane_height = 540;
+        ctx->font_size = 36;
+        break;
+    case FF_PROFILE_ARIB_PROFILE_C:
+        profile = ARIBCC_PROFILE_C;
+        ctx->plane_width = 320;
+        ctx->plane_height = 180;
+        ctx->font_size = 16;
+        break;
+    default:
+        av_log(avctx, AV_LOG_ERROR, "Unknown or unsupported profile set.\n");
+        return AVERROR(EINVAL);
+    }
+    /* determine BorderStyle of ASS header */
+    if (ctx->ignore_background)
+        ctx->border_style = 1;
+    else
+        ctx->border_style = 4;
+    ctx->charstyle = ARIBCC_CHARSTYLE_DEFAULT;
+    if (ctx->force_stroke_text || ctx->ignore_background)
+        ctx->charstyle |= ARIBCC_CHARSTYLE_STROKE;
+
+    encode_scheme = ARIBCC_ENCODING_SCHEME_AUTO;
+    if (avctx->sub_charenc_mode == FF_SUB_CHARENC_MODE_DO_NOTHING)
```

Comment:
This is meant to control libavcodec's built-in encoding conversion infrastructure. If you want to let callers control the encoding libaribcaption uses, you should provide an `AVOption` defaulting to `ARIBCC_ENCODING_SCHEME_AUTO`.

Reply:
Encoding depends on broadcast source (i.e., country).
Most of cases, it should be determined automatically.
It is last resort when it cannot be determined automatically.
If `sub_charenc_mode` is not suitable means, I'll add one in `AVOption`.

Reply:
Tentatively, I add `-encoding_scheme` option and specify it as same value as `aribcc_encoding_scheme_t`:
* auto
* jis
* utf8
* latin

I concern the option name might be too generic and it may conflct with other codecs.

Reply:
Nice! If you want to be more specific, you can use `arib_text_encoding` or the like, but I don't think it's a big deal either way.

Reply:
I determined the option name as `-caption_encoding`.
This should not be confused.